### PR TITLE
Merge 2.10.x + PR 6113 to 2.11.x [ci: last-only]

### PR DESCRIPTION
**Pull request by retronym**
This brings in backports of basic support for Java 9.

When running on Java 9, the command line `scala` and `scalac` runner scripts must be run with `-nobootcp` to allow the compiler access to `javax.tools.ToolProvider`, which is the basis of our implementation the platform classpath.

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/Implicits.scala [hunk 1]
```diff
@@ -1025,6 +1025,7 @@ trait Implicits {
                 getParts(bts(i))
                 i += 1
               }
+              Array("") :+ "2"
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Implicits.scala]: stray line?

File: src/compiler/scala/tools/util/PathResolver.scala [hunk 2]
```diff
@@ -268,9 +269,20 @@ abstract class PathResolverBase[BaseClassPathType <: ClassFileLookup[AbstractFil
 
     import classPathFactory._
 
+    private def javaBootClasspath = {
+      val cp = classesInPath(javaBootClassPath)
+      val okay = cp.exists(_.findClass("java/lang/Object.class").isDefined)
```
Comment (adriaanm) [inline src/compiler/scala/tools/util/PathResolver.scala]: can we avoid this scan unless we're under java 9?

**Review by adriaanm — COMMENTED** [review]
Minor comments. I forget the TODO you mentioned yesterday re: documenting bootcp or something?

File: src/compiler/scala/tools/util/PathResolver.scala (hunk 2 above)
Comment (retronym) [reply src/compiler/scala/tools/util/PathResolver.scala]: done

**Comment by retronym** [comment]
> Minor comments. I forget the TODO you mentioned yesterday re: documenting bootcp or something?

I've noted the limitation in the PR description.

**Comment by retronym** [comment]
I've dropped the commit that used `jrt://`/NIO for the platform classpath impl: I'd forgotten that 2.11 also needs to build on Java 6!

**Pull request merged**